sql/src/main/java/io/crate/lucene/ArrayLengthQuery.java

```diff
@@ -0,0 +1,413 @@
+            case GtOperator.NAME:
+                return x -> x > cmpValue;
+
+            case GteOperator.NAME:
+                return x -> x >= cmpValue;
+
+            case EqOperator.NAME:
+                return x -> x == cmpValue;
+
+            default:
+                throw new IllegalArgumentException("Unknown comparison function: " + cmpFuncName);
+        }
+    }
+
+    private static IntUnaryOperator getNumTermsPerDocFunction(LeafReader reader, Reference ref) {
+        DataType elementType = CollectionType.unnest(ref.valueType());
+        switch (elementType.id()) {
+            case BooleanType.ID:
+                return numValuesPerDocForBoolean(reader, ref.column());
+
+            case ByteType.ID:
+            case ShortType.ID:
+            case IntegerType.ID:
+            case LongType.ID:
+            case TimestampType.ID:
+            case FloatType.ID:
+            case DoubleType.ID:
+            case GeoPointType.ID:
+                return numValuesPerDocForSortedNumeric(reader, ref.column());
+
+            case StringType.ID:
+                return numValuesPerDocForString(reader, ref.column());
+
+            case IpType.ID:
+                return numValuesPerDocForIP(reader, ref.column());
+
+            default:
+                throw new UnsupportedOperationException("NYI: " + elementType);
+        }
+    }
+
+    private static IntUnaryOperator numValuesPerDocForIP(LeafReader reader, ColumnIdent column) {
+        SortedSetDocValues docValues;
+        try {
+            docValues = reader.getSortedSetDocValues(column.fqn());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return doc -> {
+            try {
+                if (docValues.advanceExact(doc)) {
+                    int c = 1;
+                    while (docValues.nextOrd() != SortedSetDocValues.NO_MORE_ORDS) {
+                        c++;
+                    }
+                    return c;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            return 0;
+        };
+    }
+
+    private static IntUnaryOperator numValuesPerDocForBoolean(LeafReader reader, ColumnIdent column) {
+        SortedBinaryDocValues docValues;
+        try {
+            docValues = FieldData.toString(DocValues.getSortedNumeric(reader, column.fqn()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return doc -> {
+            try {
+                return docValues.advanceExact(doc) ? docValues.docValueCount() : 0;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        };
+    }
+
+    private static IntUnaryOperator numValuesPerDocForString(LeafReader reader, ColumnIdent column) {
+        SortedBinaryDocValues docValues;
+        try {
+            docValues = FieldData.toString(DocValues.getSortedSet(reader, column.fqn()));
```

Comment:
same here, I think we can skip wrapping into `toString`

Reply:
`SortedSetDocValues` doesn't have a `docValueCount()` method. So I'd have to to loop over `nextOrd` - which is pretty much what `toString` does:

```
    public static SortedBinaryDocValues toString(final SortedSetDocValues values) {
        return new SortedBinaryDocValues() {
            private int count = 0;

            @Override
            public boolean advanceExact(int doc) throws IOException {
                if (values.advanceExact(doc) == false) {
                    return false;
                }
                for (int i = 0; ; ++i) {
                    if (values.nextOrd() == SortedSetDocValues.NO_MORE_ORDS) {
                        count = i;
                        break;
                    }
                }
                // reset the iterator on the current doc
                boolean advanced = values.advanceExact(doc);
                assert advanced;
                return true;
            }

            @Override
            public int docValueCount() {
                return count;
            }

            @Override
            public BytesRef nextValue() throws IOException {
                return values.lookupOrd(values.nextOrd());
            }

        };
    }
```

Reply:
ah sure didn't see that.